pkg/devices/common.go

```diff
@@ -13,350 +16,474 @@ import (
-		return nil, nil // No reading
+	for z := 0; z < len(sorted); z++ {
+		log.Debugf("MapBulkRead sorted[%v]: %#v", z, sorted[z])
 	}
 
-	raw := block.Results[startOffset:endOffset]
-	return UnpackReading(output, device, raw)
-}
-
-// UnpackCoilReading gets a coil reading (true / false) for the specified device from the
-// bulk read block results bytes.
-func UnpackCoilReading(output *output.Output, block *ReadBlock, device *ModbusDevice) (*output.Reading, error) {
-	bitIndex := device.Config.Address - block.StartRegister
-	byteIndex := bitIndex / 8
-	bitIndex = bitIndex % 8
+	for i := 0; i < len(sorted); i++ {
+		// Create the key for this device from the device data.
+		device := sortedDevices[sorted[i]]
+		log.Debugf("--- next synse device: %v", device)
+		var deviceData config.ModbusDeviceData
+		err = mapstructure.Decode(device.Data, &deviceData)
+		if err != nil {
+			// Hard failure on configuration issue.
+			log.Errorf(
+				"MapBulkRead failed parsing device.Data, device at:[%v], device: %#v",
+				i, device)
+			return nil, keyOrder, err
+		}
 
-	l := log.WithFields(log.Fields{
-		"address":       device.Config.Address,
-		"startRegister": block.StartRegister,
-		"resultsLen":    len(block.Results),
-		"bitIndex":      bitIndex,
-		"byteIndex":     byteIndex,
-	})
-	l.Debug("unpacking coil reading")
+		key := ModbusBulkReadKey{
+			Host:                 deviceData.Host,
+			Port:                 deviceData.Port,
+			Timeout:              deviceData.Timeout,
+			FailOnError:          deviceData.FailOnError,
+			MaximumRegisterCount: MaximumRegisterCount,
+		}
+		log.Debugf("Created key: %#v", key)
 
-	if int(byteIndex) >= len(block.Results) {
-		l.Error("failed to get coil data: index out of bounds")
-		if device.Config.FailOnError {
-			return nil, errors.New("failed to get coil data: index out of bounds")
+		// Find out if the key is in the map.
+		keyValues, keyPresent := bulkReadMap[key]
+		if keyPresent {
+			log.Debugf("key is already in the map")
+		} else {
+			log.Debugf("key is not in the map")
 		}
-		l.Debug("not failing on error, returning nil device reading")
-		return nil, nil // No reading
-	}
 
-	coilByte := block.Results[byteIndex]
-	mask := byte(0x01 << bitIndex)
-	coilState := (coilByte & mask) != 0
+		log.Debugf("len(keyValues): %v", len(keyValues))
+
+		deviceDataAddress := deviceData.Address
+		deviceDataWidth := deviceData.Width
+
+		log.Debugf("deviceDataAddress: 0x%04x", deviceDataAddress)
+		log.Debugf("deviceDataWidth: %d", deviceDataWidth)
+
+		// Insert.
+		// If the key is not present, this is a simple insert to the map.
+		if !keyPresent {
+			log.Debugf("Key not present.")
+			modbusBulkRead, err := NewModbusBulkRead(device, deviceDataAddress, deviceDataWidth, isCoil)
+			if err != nil {
+				return nil, keyOrder, err
+			}
+			log.Debugf("appending modbusBulkRead: %#v", modbusBulkRead)
+			bulkReadMap[key] = append(bulkReadMap[key], modbusBulkRead)
```

Comment:
~I'm  probably missing something, but is this right? If `!keyPresent`, doesn't that mean that the `key` is not in `bulkReadMap`,  so when you  do `bulkReadMap[key]`, it would evaluate to nil?  so `append(bulkReadMap[key], modbusBulkRead)`  would be appending `modbusBulkRead`  to `nil`?~

~Maybe an empty slice  is set for the key  somewhere and  I just missed it. Just bringing it up in case.~

Reply:
Or maybe its a behavior of Go that I didn't expect where the `nil` it  returns is a nil `[]*ModbusBulkRead`  since thats the declared value type and doing an  append on a nil list instantiates it and populates it

EDIT: oh, yeah thats exactly what it is https://play.golang.org/p/t4kNzgrAm1H 

disregard these comments